ComponentInput.type 只有一个粗粒度 number 臂,无法表达整数/正数约束 —— 发布面比契约宽,checkType 对 spec 会拒的值零诊断

发现于 #4668 的实现(PR 见下),**不在该 PR 处理** —— 那一单的面是把五个 GA 键发布为 `inputs`。这一条是发布面本身的表达力缺口,与 #3832(`type` 无法表达联合)同族、但不同维:那条是「臂不够多」,这条是「臂不够细」。

## 机制

`ComponentInputControlType`(`packages/types/src/base.ts:362-373`)的数值种类只有一个 `'number'`,`ComponentInput` 也没有任何 integer / min / max 槽位。`sdui-parser` 的 `checkType`(`packages/sdui-parser/src/validate.ts`)对该臂只问「是不是 number」。

而契约比这细。实测 `@objectstack/spec@17.0.0` 的 `ComponentPropsMap['page:header']`:

```
maxVisible        optional -> number
  checks: [{format:"safeint"},{check:"greater_than",value:0,inclusive:false}]
mobileMaxVisible  同上
```

实测判定(`PageHeaderProps.safeParse`):

| 值 | spec | objectui manifest 门(`type: 'number'`) | 渲染器 `readMax` |
|---|---|---|---|
| `1` / `2` / `3` / `100` | OK | 无诊断 | 生效 |
| `0` | **REJECT** | **无诊断** | 生效 → 0 个内联按钮,全进溢出菜单 |
| `-1` | **REJECT** | **无诊断** | 落回默认(`v >= 0` 不成立) |
| `1.5` | **REJECT** | **无诊断** | 生效 → `Math.floor` 成 1 |

## 三个权威,两个答案

作者写 `maxVisible: 0`:objectui 这边一路静默(manifest 白名单绿、`checkType` 绿),渲染器照它渲染,而上游 `os validate` / `os build` 按值拒掉整份元数据。这正是 parity 门自己文件头引的那句 objectstack#5435「platform authority must not point at keys its own gate rejects」的**值维度版本** —— 键名对上了,值域没对上。

#4668 的处置是把约束写进 description(「A positive integer — the contract rejects 0 and fractional values」),那是当下唯一能表达它的地方,但 description 不是门。

## 附带一半:渲染器比契约更宽

`readMax`(`packages/components/src/renderers/layout/containers.tsx`,`page:header` 的内联/溢出切分处)是

```
typeof v === 'number' && Number.isFinite(v) && v >= 0 ? Math.floor(v) : undefined
```

—— 接受 `0`、接受小数并下取整,两者 spec 都拒。仓内没有任何生产者写这些值(grep 无),所以今天是**未被行使的漂移**而不是活方言;但它与上面同向:三层里最松的一层决定了实际行为。

## 可能的处置(供分诊,未实施)

1. **给 `ComponentInput` 加数值约束槽**(如 `integer?: boolean` / `min?` / `max?`),`checkType` 读它 —— 判据与被判定的事实对齐,形状与 #3832 给 `type` 加数组臂时的做法一致(维护者 2026-08-09 裁决方向 a)。代价:`ComponentInput` 变宽,且要回答「约束是抄 spec 还是独立声明」——抄就有漂移风险,独立声明就有两份真相。
2. **反过来:让 `checkType` 对数值臂去问 spec** —— 不给 `ComponentInput` 加槽,而是让门在有 `ComponentPropsMap` 条目时直接用 spec 的 Zod 成员判值。判据只有一份,但把 sdui-parser 绑到 spec 上(它今天只吃 manifest,不吃 spec)。
3. **只收紧渲染器**(`readMax` 改成 `> 0` 且拒小数),让渲染器与契约一致,发布面的粗粒度照旧。最小、独立可落,但不解决「objectui 侧对非法值零诊断」。
4. **不动,承认 description 是唯一表达处** —— 记录取舍,把 spec 作为值的唯一裁判。

倾向 1 或 2(判据与事实对齐;2 更彻底也更侵入),3 可以独立先落。**这是发布契约的形状决定,值得维护者拍一下,不建议实现者自选。**

## 退出条件

- 上述任一处置落地;或
- 明确裁定「粗粒度 `number` + description 就是发布面的表达上限」,并把该取舍写进 `ComponentInput.type` 的 doc comment(那里已经为 #3832 写过同类说明)。

参考位置:`packages/types/src/base.ts:362-373`(`ComponentInputControlType`)、`packages/sdui-parser/src/validate.ts`(`armAccepts` / `checkType`)、`packages/components/src/renderers/layout/containers.tsx`(`readMax` 与 `page:header` 的两条 inputs)、`apps/console/src/__tests__/registry-inputs-spec-parity.test.ts` 文件头 LIMIT 段(已写明本门只比顶层键名、类型比较「is not, and is nobody's check yet」)。

<!-- os-dev-report -->

```json
{
  "issue": 5006,
  "status": "done",
  "branch": "claude/issue-5006-readmax-contract-tightening",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5298",
  "premise_still_valid": true,
  "summary": "Both ruled deliverables landed in ONE draft PR with `Fixes #5006`. (1) `readMax` at packages/components/src/renderers/layout/containers.tsx now enforces the contract: `Number.isSafeInteger(v) && v > 0`, replacing `Number.isFinite(v) && v >= 0 ? Math.floor(v)`. A contract-rejected value falls back to the documented default (3 desktop / 1 mobile) instead of taking effect. (2) `ComponentInput.type`'s doc comment at packages/types/src/base.ts records the ruled trade alongside the existing #3832 note: the coarse `number` arm plus `description` IS the publication face's expression ceiling today, spec is the sole judge of values, and dispositions 1 and 2 are deferred with the ruling's named reopen condition quoted. Dispositions 1 and 2 were NOT taken. SERIAL CONSTRAINTS: all three cleared by reading, not by elapsed time -- #5261 holds .github/workflows/ + scripts/, #5264 holds packages/plugin-dashboard/src/**, #5280 holds one plugin-dashboard test file. All disjoint from both targets. RE-DERIVATION on origin/main 95cf24168 (not round 12's 92b327a67): all three authorities still disagreed exactly as filed, and readMax had not been tightened by anyone else. TWO REFINEMENTS TO THE RECORD, neither changing the conclusion. (a) `Number.isSafeInteger`, NOT `Number.isInteger`, is the exact translation of spec's `{format:'safeint'}` -- I probed the edges and spec REJECTS `MAX_SAFE_INTEGER + 2` and `1e21` ('Too big: expected int to be <=9007199254740991') while `Number.isInteger` returns true for both. Using `isInteger` would have left a fresh, smaller version of the very gap this card exists to close. (b) ROUND 12'S PRODUCER SWEEP WAS SLIGHTLY WRONG and the correction is worth recording because the claim comment carried it forward as established fact: the sweep reported 'only positive integers (1, 2, 4, 5, 10)', but there IS one `maxVisible: 0` in the tree, at packages/components/src/renderers/action/__tests__/action-record-predicate-root.test.tsx:437. It belongs to `action:bar`, NOT to readMax -- exactly the different-reader trap the dispatch warned about -- so the ruling's premise still holds for page:header and the tightening breaks nothing. I verified the separation rather than assuming it: `action:bar` reads `schema.maxVisible ?? 3` at action-bar.tsx:229-231 through a wholly separate reader, and it is NOT in `ComponentPropsMap` at all, so no contract constrains it. I also removed the now-false phrase 'the same contract as action:bar' from the readMax comment and corrected a stale claim in the same file's `inputs` block that read '`readMax` here is laxer than that' -- this PR falsifies it.",
  "tests": "All gates run at FINAL commit 616499ba7 with a clean tree (`git status --porcelain` empty, `git diff HEAD` empty). Canonical invocation used throughout -- from the REPO ROOT with no `--`, per scripts/vitest-invocation-guard.mjs, which refuses package-cwd runs and `--`-swallowed path filters as false greens. (1) TARGETED: `pnpm exec vitest run packages/components/src/__tests__/page-header-actions.test.tsx` -> 'Test Files 1 passed (1) / Tests 43 passed (43)' (37 pre-existing + 6 new). (2) AFFECTED PACKAGES: `pnpm exec vitest run packages/components packages/types` -> 'Test Files 202 passed (202) / Tests 1932 passed (1932)'. (3) DOWNSTREAM PARITY GATES that read this renderer's `inputs`: `pnpm exec vitest run apps/console/src/__tests__/registry-inputs-spec-parity.test.ts apps/console/src/__tests__/ga-honoured-inputs-author-reach.test.ts` -> 'Test Files 2 passed (2) / Tests 69 passed (69)'. (4) `pnpm --filter @object-ui/types --filter @object-ui/components type-check` -> both 'Done', exit 0 (this repo spells it `type-check`, HYPHENATED -- the misspelling trap that exits 0 having run nothing; I confirmed the script names from package.json first and the output echoed both). (5) lint on the same two packages -> '0 errors, 891 warnings', all warnings pre-existing and package-wide; the `v: any` signature is unchanged by this PR. (6) `check:control-bytes` OK (4685 tracked files), `check:doc-types` OK, `check-changeset-no-major` OK. Also self-scanned all four changed files with `grep -naP '[\\x00-\\x08\\x0b\\x0c\\x0e-\\x1f\\x7f]'` -> clean. SPEC RE-DERIVATION, with the counter-probe the dispatch demanded: probing the `@objectstack/spec` package ROOT gives 125 exports with `ComponentPropsMap` ABSENT -- I counter-probed by filtering the root's export names for /Component|Props/ and got an empty list, confirming the absence is real rather than a failed module load; the member lives on the `@objectstack/spec/ui` subpath, which exports 218 names, and its map holds 37 keys with `page:header` present (counter-probed against sibling keys page:tabs, page:card, page:footer). Live safeParse there: 1/3/100/MAX_SAFE_INTEGER OK; 0 and -1 REJECT 'Too small: expected number to be >0'; 1.5 and 2.5 REJECT 'Invalid input: expected int, received number'; MAX_SAFE_INTEGER+2 and 1e21 REJECT 'Too big'; Infinity/NaN REJECT. ABLATION -- REBUILD STATEMENT: NO build artifact sits between the edit and the thing under test, so no rebuild was required for the ablation to be real. The root vitest.config.mts aliases `@object-ui/components` to packages/components/src (and the package's own vite.config self-references it to ./src), so resolution never reaches the package `exports` -> `dist`; this is the vitest-alias case, not the dist-resolved case that needs a dist preflight. I stated predictions BEFORE running: 4 red / 2 green-both-legs, 4 failed + 39 passed of 43, naming the first failing assertion in each. OBSERVED EXACTLY THAT: 'Tests 4 failed | 39 passed (43)' with `maxVisible: 0` -> 'Unable to find an accessible element with the role \"button\" and name `/Action A/i`'; `1.5` -> same for `/Action B/i` (old code floored to 1); `MAX_SAFE_INTEGER + 2` -> 'AssertionError: expected <button ...> to be null' (old code passed the huge budget through so nothing overflowed); `properties.maxVisible: 0` -> same as case 1. Direction was the ordinary reddening one. The fix was COMMITTED BEFORE the ablation, so the restore was `git checkout HEAD -- <path>`, and I proved byte-identical restoration (`git status --porcelain` and `git diff HEAD --stat` both empty) and re-ran the RESTORE LEG to green: 'Tests 43 passed (43)'. HONEST ACCOUNTING: 2 of my 6 new tests are GREEN ON BOTH LEGS and pin nothing about this change -- a negative maxVisible already fell back (the old guard was `v >= 0`) and a positive safe integer was already honoured. They are labelled in the test file and the PR as regression guards only and are NOT counted as evidence for the fix. The 4 that discriminate are the ones listed above.",
  "open_questions": [],
  "out_of_scope_findings": []
}
```

**No out-of-scope cards filed, and the one candidate is recorded here rather than left silent.** `action:bar` validates `maxVisible` not at all (`schema.maxVisible ?? 3`), so a negative value reaches `slice()`. I chose not to file it after measuring rather than on a size judgement: `action:bar` has **no `ComponentPropsMap` entry**, so unlike `page:header` there is no contract for it to diverge from, and its `maxVisible: 0` fixture shows the lax read is a deliberately exercised idiom there. A negative value merely moves the last action into the overflow menu — every action stays reachable, so there is no defect to report. Filing it would also have amounted to asking for disposition-1/2-shaped work on a sibling surface, which the ruling deferred. Flagging it for the PM to overrule if that reasoning is wrong.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/page-header-maxvisible-contract-domain-5006.md
+++ b/.changeset/page-header-maxvisible-contract-domain-5006.md
@@ -1,0 +1,39 @@
+---
+'@object-ui/components': patch
+'@object-ui/types': patch
+---
+
+`page:header`'s `maxVisible` / `mobileMaxVisible` now honour the contract's value domain instead of a laxer renderer tolerance.
+
+Three authorities gave two answers for the same value (objectui#5006). Measured on
+`ComponentPropsMap['page:header']` at `@objectstack/spec@17.0.0` — the member lives
+on the `@objectstack/spec/ui` subpath, not the package root — both keys are a
+POSITIVE SAFE INTEGER (`{format:'safeint'}` plus
+`{check:'greater_than',value:0,inclusive:false}`). Spec rejects `0`, `-1`, `1.5`
+and anything past `Number.MAX_SAFE_INTEGER`. objectui's manifest gate and
+`sdui-parser`'s `checkType` said nothing about any of them, and the renderer's
+`readMax` was looser still: it accepted `0` and floored fractions. So the loosest
+of the three layers decided what shipped on screen, while `os validate` / `os build`
+rejected the very same metadata outright.
+
+`readMax` now accepts only what the contract accepts. `Number.isSafeInteger(v) && v > 0`
+is the exact translation of `safeint`, not an approximation — plain `Number.isInteger`
+would admit `2**53 + 2` and `1e21`, which spec rejects.
+
+Behaviour change, stated because this NARROWS the renderer's accept set rather than
+only fixing a fault: a contract-rejected value no longer takes effect and falls back
+to the documented default (3 desktop / 1 mobile). Concretely, `maxVisible: 0` used to
+render zero inline buttons and sweep every action into the overflow menu, and
+`maxVisible: 1.5` used to be floored to `1`; both now render the default 3-inline
+split. This is a narrowing *toward* an already-published contract — no in-tree
+producer writes a rejected value, so nothing in the repo changes behaviour. Both
+schema-level and `properties.*` spellings go through the one reader. `action:bar`'s
+`maxVisible` is an unrelated reader with no `ComponentPropsMap` entry and is
+deliberately untouched.
+
+`ComponentInput.type`'s doc comment now records the trade the ruling fixed in place
+(maintainer, 2026-08-17): the coarse `number` arm plus `description` is the
+publication face's expression ceiling today, and spec is the sole judge of values.
+Giving `ComponentInput` real constraint slots, and binding `checkType` to spec, were
+both deferred with a named reopen condition — a measured case of an author shipping
+a spec-rejected value that objectui's silence let through.

--- a/packages/components/src/__tests__/page-header-actions.test.tsx
+++ b/packages/components/src/__tests__/page-header-actions.test.tsx
@@ -429,6 +429,107 @@ describe('PageHeaderRenderer — inline/overflow split (objectui#2361)', () => {
   });
 });
 
+// objectui#5006 — the `maxVisible` / `mobileMaxVisible` VALUE DOMAIN.
+//
+// Three authorities used to give two answers for the same value. Measured on
+// `ComponentPropsMap['page:header']` at @objectstack/spec 17.0.0 (the member
+// lives on the `@objectstack/spec/ui` subpath, NOT the package root), both keys
+// are a positive safe integer — `{format:'safeint'}` plus
+// `{check:'greater_than',value:0,inclusive:false}`:
+//
+//   value | spec                                  | old readMax
+//   ------+---------------------------------------+---------------------------
+//   3     | OK                                    | 3
+//   0     | REJECT "expected number to be >0"     | 0  (all actions overflowed)
+//   -1    | REJECT "expected number to be >0"     | undefined -> default
+//   1.5   | REJECT "expected int, received number"| 1  (floored)
+//   2^53+2| REJECT "expected int to be <=2^53-1"  | 2^53+2 (nothing overflowed)
+//
+// The renderer was the loosest of the three, so a value `os validate` rejects
+// outright still changed what shipped on screen. `readMax` now enforces the
+// contract and a rejected value falls back to the documented default (3 desktop
+// / 1 mobile) instead of taking effect.
+//
+// ⚠️ Scope note for future readers: `readMax` is `page:header`-only and has
+// exactly two call sites. `action:bar` reads `schema.maxVisible ?? 3` through a
+// wholly separate reader (`renderers/action/action-bar.tsx`) and is NOT in
+// `ComponentPropsMap` at all, so its `maxVisible: 0` fixtures are legitimate and
+// untouched by this suite. The notification stack's `maxVisible` is a third,
+// unrelated reader (provider config).
+describe('PageHeaderRenderer — maxVisible contract domain (objectui#5006)', () => {
+  const fourActions = [
+    { name: 'a', locations: ['record_header'], label: 'Action A' },
+    { name: 'b', locations: ['record_header'], label: 'Action B' },
+    { name: 'c', locations: ['record_header'], label: 'Action C' },
+    { name: 'd', locations: ['record_header'], label: 'Action D' },
+  ];
+
+  // Asserts the DEFAULT-3 split: A/B/C inline, D folded into the overflow menu.
+  // That is what a contract-rejected override must fall back to.
+  function expectDefaultThreeSplit() {
+    expect(screen.getByRole('button', { name: /Action A/i })).toBeTruthy();
+    expect(screen.getByRole('button', { name: /Action B/i })).toBeTruthy();
+    expect(screen.getByRole('button', { name: /Action C/i })).toBeTruthy();
+    expect(screen.queryByRole('button', { name: /Action D/i })).toBeNull();
+    expect(screen.getByRole('button', { name: /More actions/i })).toBeTruthy();
+  }
+
+  it('rejects maxVisible: 0 and falls back to the default (spec rejects 0)', () => {
+    // Pre-fix this rendered ZERO inline buttons and swept every action into the
+    // overflow menu — the most visible of the divergences, and the one an author
+    // was most likely to hit by writing "0 means unlimited".
+    renderHeader({ type: 'page:header', maxVisible: 0, actions: fourActions });
+    expectDefaultThreeSplit();
+  });
+
+  it('rejects a fractional maxVisible instead of flooring it (spec rejects 1.5)', () => {
+    // Pre-fix: Math.floor(1.5) === 1, so only Action A stayed inline. Flooring
+    // invented a value the contract never accepted.
+    renderHeader({ type: 'page:header', maxVisible: 1.5, actions: fourActions });
+    expectDefaultThreeSplit();
+  });
+
+  it('rejects a maxVisible past Number.MAX_SAFE_INTEGER (spec format is safeint)', () => {
+    // Pre-fix this was finite and >= 0, so it passed straight through and every
+    // action rendered inline with no overflow button at all. This is the case
+    // that makes Number.isSafeInteger — not Number.isInteger — the correct
+    // predicate: Number.isInteger(2**53 + 2) is true, but spec rejects it.
+    renderHeader({
+      type: 'page:header',
+      maxVisible: Number.MAX_SAFE_INTEGER + 2,
+      actions: fourActions,
+    });
+    expectDefaultThreeSplit();
+  });
+
+  it('applies the same domain to the properties.* spelling (spec bridge variant)', () => {
+    // Both spellings flow through the one `readMax`, so the bridge variant must
+    // not be a way around the contract.
+    renderHeader({
+      type: 'page:header',
+      properties: { maxVisible: 0, actions: fourActions },
+    });
+    expectDefaultThreeSplit();
+  });
+
+  // ⚠️ The two tests below are GREEN BOTH BEFORE AND AFTER the fix — they pin
+  // no part of this change and are recorded as regression guards only. A
+  // negative override already fell back (the old guard was `v >= 0`), and a
+  // positive safe integer was already honoured. They are kept so the accepted
+  // and rejected halves of the domain are both stated in one place.
+  it('regression guard (green pre-fix too): a negative maxVisible falls back', () => {
+    renderHeader({ type: 'page:header', maxVisible: -1, actions: fourActions });
+    expectDefaultThreeSplit();
+  });
+
+  it('regression guard (green pre-fix too): a positive safe integer is honoured', () => {
+    renderHeader({ type: 'page:header', maxVisible: 2, actions: fourActions });
+    expect(screen.getByRole('button', { name: /Action B/i })).toBeTruthy();
+    expect(screen.queryByRole('button', { name: /Action C/i })).toBeNull();
+    expect(screen.getByRole('button', { name: /More actions/i })).toBeTruthy();
+  });
+});
+
 // objectui#3391 — record dispatch shape. A `record_header` / `record_more`
 // action must reach the runtime in the SAME shape ObjectGrid row actions and
 // DeclaredActionsBar dispatch: the record stashed under `params._rowRecord`

--- a/packages/components/src/renderers/layout/containers.tsx
+++ b/packages/components/src/renderers/layout/containers.tsx
@@ -1400,12 +1400,34 @@ const PageHeaderRenderer: React.FC<any> = ({ schema, className, ...props }) => {
     //      (see the headerActions memo above), so the first `maxVisible`
     //      claim the inline slots.
     // `maxVisible` / `mobileMaxVisible` are overridable on the page:header
-    // schema and default to 3 / 1 — the same contract as action:bar. This
-    // still keeps the header from drowning in 4–5 buttons on every record
-    // page, while letting multi-action objects surface several primary
-    // buttons at once.
+    // schema and default to 3 / 1. This still keeps the header from drowning
+    // in 4–5 buttons on every record page, while letting multi-action objects
+    // surface several primary buttons at once.
+    //
+    // `readMax` enforces the SPEC's value domain rather than a laxer renderer
+    // tolerance (objectui#5006). Measured on `ComponentPropsMap['page:header']`
+    // at @objectstack/spec 17.0.0, both keys are a POSITIVE SAFE INTEGER
+    // (`{format:'safeint'}` + `{check:'greater_than',value:0,inclusive:false}`),
+    // so `0`, `-1`, `1.5` and anything past `Number.MAX_SAFE_INTEGER` are all
+    // rejected by the contract. This reader used to be the loosest of the three
+    // authorities — it accepted `0` and floored fractions — so a value that
+    // `os validate` / `os build` rejects outright still changed what rendered
+    // here, and the loosest layer decided behaviour. A contract-rejected value
+    // now falls back to the default instead of taking effect.
+    //
+    // `Number.isSafeInteger` is the exact translation of `safeint`, not an
+    // approximation: plain `Number.isInteger` would admit `2**53 + 2` and
+    // `1e21`, both of which spec rejects (`Too big: expected int to be
+    // <= 9007199254740991`). It also subsumes the old `Number.isFinite` guard,
+    // since Infinity and NaN are not safe integers.
+    //
+    // ⛔ Do NOT re-lax this into a tolerant read. objectui's authoring-time
+    // gates stay coarse by ruling (maintainer, 2026-08-17): `ComponentInput`
+    // keeps its single `number` arm and `checkType` is not bound to spec, so
+    // SPEC IS THE SOLE JUDGE OF VALUES — which makes agreeing with it this
+    // renderer's job. Rejections are pinned in `page-header-actions.test.tsx`.
     const readMax = (v: any): number | undefined =>
-      typeof v === 'number' && Number.isFinite(v) && v >= 0 ? Math.floor(v) : undefined;
+      typeof v === 'number' && Number.isSafeInteger(v) && v > 0 ? v : undefined;
     const maxVisible = isMobile
       ? (readMax(schema?.mobileMaxVisible ?? schema?.properties?.mobileMaxVisible) ?? 1)
       : (readMax(schema?.maxVisible ?? schema?.properties?.maxVisible) ?? 3);
@@ -1748,10 +1770,13 @@ ComponentRegistry.register('header', PageHeaderRenderer, {
     // `ComponentPropsMap['page:header']` on the installed GA pin, both are
     // `z.number()` constrained to a POSITIVE SAFE INTEGER — `0`, `-1` and `2.5`
     // are all rejected by value, which is why the descriptions say so instead
-    // of leaving an author to discover it from a parse error. (`readMax` here is
-    // laxer than that — it accepts `0` and floors fractions — but a renderer
-    // tolerance is not an authoring surface, and the contract is what `inputs`
-    // publishes.)
+    // of leaving an author to discover it from a parse error. (`readMax` at the
+    // inline/overflow split above now enforces exactly that domain — objectui#5006
+    // closed the gap where this renderer accepted `0` and floored fractions while
+    // the contract rejected both. The coarse `number` arm here still cannot
+    // express the domain: `ComponentInput.type` has no integer/min/max slot, so
+    // `description` stays the only authoring-time expression of it, and spec
+    // stays the sole judge of values.)
     { name: 'maxVisible', type: 'number', label: 'Max Inline Actions', defaultValue: 3, description: 'How many header actions render as inline buttons on desktop before the rest fold into the overflow menu. A positive integer — the contract rejects 0 and fractional values. Two kinds of action are routed to the overflow menu regardless of this budget and never occupy an inline slot: an action whose locations declare record_more without record_header, and any action with component action:menu.' },
     { name: 'mobileMaxVisible', type: 'number', label: 'Max Inline Actions (Mobile)', defaultValue: 1, description: 'The same inline-button budget on mobile viewports, where horizontal room is scarce. A positive integer, defaulting to 1; the desktop half is Max Inline Actions, and the overflow routing rules stated there apply unchanged.' },
   ],

--- a/packages/types/src/base.ts
+++ b/packages/types/src/base.ts
@@ -410,6 +410,35 @@ export interface ComponentInput {
    * one-element array back to the bare string, so the published
    * `sdui.manifest.json` gains arrays only where a union was really declared.
    *
+   * ## The arms are COARSE KINDS — and that is the ceiling (objectui#5006)
+   *
+   * An arm names a value's *kind*, never its *domain*. `'number'` is the only
+   * numeric arm and `ComponentInput` has no `integer` / `min` / `max` slot to
+   * pair with it, so a key whose contract is narrower than "some number" has
+   * no way to say so here. Worked example — `page:header.maxVisible`, whose
+   * spec type is a POSITIVE SAFE INTEGER: `@objectstack/spec` rejects `0`,
+   * `-1` and `1.5`, while this declaration's `'number'` arm admits all three
+   * and `checkType` raises no diagnostic on any of them.
+   *
+   * **Ruled (maintainer, 2026-08-17): the coarse arm plus `description` IS the
+   * publication face's expression ceiling today, and SPEC IS THE SOLE JUDGE OF
+   * VALUES.** So spell the domain out in `description`, in the author's own
+   * terms ("A positive integer — the contract rejects 0 and fractional
+   * values"), and let `os validate` / `os build` be the gate that enforces it.
+   * `description` documents, it does not check: objectui deliberately raises
+   * no authoring-time diagnostic on an out-of-domain value. A *renderer*
+   * reading such a key should therefore agree with spec rather than tolerate
+   * what spec rejects — `page:header`'s `readMax` was tightened to exactly the
+   * contract for this reason, so the loosest layer stops deciding behaviour.
+   *
+   * Two directions were **deferred**, not rejected on merit: giving
+   * `ComponentInput` real constraint slots (two sources of truth, free to
+   * drift), and binding `checkType` to spec's Zod member when a
+   * `ComponentPropsMap` entry exists (one truth, but couples `sdui-parser` to
+   * spec). The ruling names the reopen condition: **a measured case of an
+   * author — human or agent — shipping a spec-rejected value that objectui's
+   * silence let through.** Until that is measured, do not widen this field.
+   *
    * @example 'string'
    * @example ['string', 'object']   // a string or an inline translation map
    * @example ['string', 'number']   // element:text_input.defaultValue

